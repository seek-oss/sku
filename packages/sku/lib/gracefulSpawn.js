const { promisify } = require('util');
const spawn = require('cross-spawn');
const treeKillAsync = promisify(require('tree-kill'));
const onDeath = require('death');

module.exports = (...args) => {
  const childProcess = spawn(...args);

  childProcess.kill = async (signal) => {
    await treeKillAsync(childProcess.pid, signal);

    // Needs a bit more time, for some reason :(
    // If we don't give it a bit of breathing room,
    // Jest complains about handles being left open.
    await new Promise((resolve) => setTimeout(resolve, 100));
  };

  onDeath((signal) => {
    childProcess.kill(signal);
  });

  return childProcess;
};
