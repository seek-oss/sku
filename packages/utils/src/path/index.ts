export { toPosixPath } from './toPosixPath.js';
