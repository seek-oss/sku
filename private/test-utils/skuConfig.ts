/**
 * A helper to make stable hashes for webpack.
 */
export const makeStableHashes = (config: Record<string, any>) => {
  // Addresses an issue where module IDs were different between local dev and CI
  config.optimization.chunkIds = 'natural';
  config.optimization.moduleIds = 'natural';

  // Disable hashes
  if (config.output.filename?.includes('[contenthash]')) {
    config.output.filename = '[name].js';
  }
  if (config.output.chunkFilename?.includes('[contenthash]')) {
    config.output.chunkFilename = '[name].chunk.js';
  }

  return config;
};

/**
 * A helper for vite to make stable hashes.
 */
export const makeStableViteHashes = () => ({
  build: {
    rollupOptions: {
      output: {
        chunkFileNames: '[name].js',
        assetFileNames: '[name][extname]',
        entryFileNames: '[name].js',
      },
    },
  },
});
