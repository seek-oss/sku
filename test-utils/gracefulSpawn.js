import { promisify } from 'node:util';
import spawn from 'cross-spawn';
import treeKill from 'tree-kill';
import onDeath from 'death';

const treeKillAsync = promisify(treeKill);

export const gracefulSpawn = (...args) => {
  try {
    const childProcess = spawn(...args);

    childProcess.kill = async (signal) => {
      await treeKillAsync(childProcess.pid, signal);

      // Needs a bit more time, for some reason :(
      // If we don't give it a bit of breathing room,
      // Jest complains about handles being left open.
      await new Promise((resolve) => setTimeout(resolve, 100));
    };

    onDeath((signal) => {
      childProcess.kill(signal);
    });

    return childProcess;
  } catch (e) {
    console.log('cross spawn issue', e);
  }
};
