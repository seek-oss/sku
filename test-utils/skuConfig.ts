export const makeStableHashes = (config: Record<string, any>) => {
  // Addresses an issue where module IDs were different between local dev and CI
  config.optimization.chunkIds = 'natural';
  config.optimization.moduleIds = 'natural';

  // Disable hashes
  if (config.output.filename?.includes('[contenthash]')) {
    config.output.filename = '[name].js';
  }
  if (config.output.chunkFilename?.includes('[contenthash]')) {
    config.output.chunkFilename = '[name].chunk.js';
  }

  return config;
};
