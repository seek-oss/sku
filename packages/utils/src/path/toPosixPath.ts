import path from 'node:path';

/**
 * Replaces all win32 path separators (\) with posix path separators (/)
 */
export const toPosixPath = (inputPath: string) =>
  inputPath.replaceAll(path.win32.sep, path.posix.sep);
