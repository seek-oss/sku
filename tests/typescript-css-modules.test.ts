import assert from 'node:assert/strict';
import path from 'path';
import { rimraf } from 'rimraf';
import {
  dirContentsToObject,
  waitForUrls,
  runSkuScriptInDir,
  getAppSnapshot,
  startAssetServer,
} from '@sku-private/test-utils';
import gracefulSpawn from '../packages/sku/lib/gracefulSpawn';

import skuConfig from '@sku-fixtures/typescript-css-modules/sku.config.ts';
import skuSsrConfig from '@sku-fixtures/typescript-css-modules/sku-ssr.config.ts';

const appDir = path.dirname(
  require.resolve('@sku-fixtures/typescript-css-modules/sku.config.ts'),
);
const distDir = path.resolve(appDir, 'dist');
const distSsrDir = path.resolve(appDir, 'dist-ssr');
const srcDir = path.resolve(appDir, 'src');

assert(skuSsrConfig.serverPort, 'sku config has serverPort');
const backendUrl = `http://localhost:${skuSsrConfig.serverPort}`;
const cssTypes = ['.less.d.ts'];

describe('typescript-css-modules', () => {
  describe('build', () => {
    assert(skuConfig.port, 'sku config has port');
    const url = `http://localhost:${skuConfig.port}`;
    let process;

    beforeAll(async () => {
      await runSkuScriptInDir('build', appDir);
      process = await runSkuScriptInDir('serve', appDir);
      await waitForUrls(url);
    });

    afterAll(async () => {
      await process.kill();
      // Clean up dist dir to prevent pollution of linted files in lint test
      await rimraf(distDir);
    });

    it('should create valid app', async () => {
      const app = await getAppSnapshot(url);
      expect(app).toMatchSnapshot();
    });

    it('should generate the expected files', async () => {
      const files = await dirContentsToObject(distDir);
      const cssTypeFiles = await dirContentsToObject(srcDir, cssTypes);
      expect({
        ...files,
        ...cssTypeFiles,
      }).toMatchSnapshot();
    });
  });

  describe('build-ssr', () => {
    let server, closeAssetServer;

    beforeAll(async () => {
      await runSkuScriptInDir('build-ssr', appDir, [
        '--config=sku-ssr.config.ts',
      ]);
      server = gracefulSpawn('node', ['server'], {
        cwd: distSsrDir,
        stdio: 'inherit',
      });
      closeAssetServer = await startAssetServer(4003, distSsrDir);
      await waitForUrls(backendUrl, 'http://localhost:4003');
    });

    afterAll(async () => {
      await server.kill();
      closeAssetServer();
      // Clean up dist-ssr dir to prevent pollution of linted files in lint test
      await rimraf(distSsrDir);
    });

    it('should create valid app', async () => {
      const app = await getAppSnapshot(backendUrl);
      expect(app).toMatchSnapshot();
    });

    it('should generate the expected files', async () => {
      const files = await dirContentsToObject(distSsrDir, ['.js', '.css']);
      const cssTypeFiles = await dirContentsToObject(srcDir, cssTypes);
      expect({
        ...files,
        ...cssTypeFiles,
      }).toMatchSnapshot();
    });
  });

  describe('start', () => {
    const devServerUrl = `http://localhost:8204`;
    let server;

    beforeAll(async () => {
      server = await runSkuScriptInDir('start', appDir);
      await waitForUrls(devServerUrl);
    });

    afterAll(async () => {
      await server.kill();
    });

    it('should start a development server', async () => {
      const snapshot = await getAppSnapshot(devServerUrl);
      expect(snapshot).toMatchSnapshot();
    });
  });

  describe('test', () => {
    let exitCode;

    beforeAll(async () => {
      const { child } = await runSkuScriptInDir('test', appDir);
      exitCode = child.exitCode;
    });

    it('should handle Less and css.js in tests', async () => {
      expect(exitCode).toEqual(0);
    });
  });

  describe('lint', () => {
    let exitCode;

    beforeAll(async () => {
      // run build first to ensure typescript declarations are generated
      await runSkuScriptInDir('build', appDir);
      const { child } = await runSkuScriptInDir('lint', appDir);
      exitCode = child.exitCode;
    });

    it('should handle tsc and eslint', async () => {
      expect(exitCode).toEqual(0);
    });
  });
});
