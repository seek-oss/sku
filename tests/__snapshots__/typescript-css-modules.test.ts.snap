// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typescript-css-modules build should create valid app 1`] = `
SCRIPTS: [
  "/static/typescript/runtime-df7fc58286eea709736e.js",
  "/static/typescript/902-6b08cd80155743245faf.js",
  "/static/typescript/main-d54504137ba99e34832b.js",
]
CSS: [
  "/static/typescript/main-cb8848ae92e1e9d08128.css",
]
SOURCE HTML: <!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      My Awesome Project
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="stylesheet"
      href="styles[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[1]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[2]"
    >
  </head>
  <body>
    <div id="app">
      <div class="eRNyDeb">
        <div
          class="MYunKdE"
          data-automation-text="true"
        >
          Hello World
        </div>
        <div>
          Render type:
          Initial
        </div>
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[0]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[1]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[2]"
    >
    </script>
  </body>
</html>
POST HYDRATE DIFFS: 
         </div>
         <div>
           Render type:
-          Initial
+          Client
         </div>
       </div>
     </div>

`;

exports[`typescript-css-modules build should generate the expected files 1`] = `
{
  "902-6b08cd80155743245faf.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "components/BlueBlock/BlueBlock.less.d.ts": "// This file is automatically generated.
// Please do not change this file!
export const border: string;
export const root: string;
",
  "components/RedBlock/RedBlock.less.d.ts": "// This file is automatically generated.
// Please do not change this file!
export const border: string;
export const root: string;
",
  "index.html": SCRIPTS: [
    "/static/typescript/runtime-df7fc58286eea709736e.js",
    "/static/typescript/902-6b08cd80155743245faf.js",
    "/static/typescript/main-d54504137ba99e34832b.js",
  ]
CSS: [
    "/static/typescript/main-cb8848ae92e1e9d08128.css",
  ]
SOURCE HTML: <!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      My Awesome Project
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="stylesheet"
      href="styles[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[1]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[2]"
    >
  </head>
  <body>
    <div id="app">
      <div class="eRNyDeb">
        <div
          class="MYunKdE"
          data-automation-text="true"
        >
          Hello World
        </div>
        <div>
          Render type:
          Initial
        </div>
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[0]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[1]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[2]"
    >
    </script>
  </body>
</html>,
  "lessStyles.less.d.ts": "// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'nested': string;
  'root': string;
}
export const cssExports: CssExports;
export default cssExports;
",
  "main-cb8848ae92e1e9d08128.css": .eRNyDeb {
  display: flex;
}
.eRNyDeb .MYunKdE {
  font-size: 32px;
}
,
  "main-d54504137ba99e34832b.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "runtime-df7fc58286eea709736e.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`typescript-css-modules build-ssr should create valid app 1`] = `
SCRIPTS: [
  "http://localhost:4003/runtime-424bd133a78585db66ac.js",
  "http://localhost:4003/675-dfa2e8f4bb72a3a26a72.js",
  "http://localhost:4003/main-bc73b281aded23a85916.js",
]
CSS: [
  "http://localhost:4003/main-cb8848ae92e1e9d08128.css",
]
SOURCE HTML: <!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      My Awesome Project
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="stylesheet"
      href="styles[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[0]"
      crossorigin="anonymous"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[1]"
      crossorigin="anonymous"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[2]"
      crossorigin="anonymous"
    >
  </head>
  <body>
    <div id="app">
      <div class="eRNyDeb">
        <div
          class="MYunKdE"
          data-automation-text="true"
        >
          Hello World
        </div>
        <div>
          Render type:
          Initial
        </div>
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
      crossorigin="anonymous"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
      crossorigin="anonymous"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[0]"
      crossorigin="anonymous"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[1]"
      crossorigin="anonymous"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[2]"
      crossorigin="anonymous"
    >
    </script>
  </body>
</html>
POST HYDRATE DIFFS: 
         </div>
         <div>
           Render type:
-          Initial
+          Client
         </div>
       </div>
     </div>

`;

exports[`typescript-css-modules build-ssr should generate the expected files 1`] = `
{
  "675-dfa2e8f4bb72a3a26a72.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "components/BlueBlock/BlueBlock.less.d.ts": "// This file is automatically generated.
// Please do not change this file!
export const border: string;
export const root: string;
",
  "components/RedBlock/RedBlock.less.d.ts": "// This file is automatically generated.
// Please do not change this file!
export const border: string;
export const root: string;
",
  "lessStyles.less.d.ts": "// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'nested': string;
  'root': string;
}
export const cssExports: CssExports;
export default cssExports;
",
  "main-bc73b281aded23a85916.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "main-cb8848ae92e1e9d08128.css": .eRNyDeb {
  display: flex;
}
.eRNyDeb .MYunKdE {
  font-size: 32px;
}
,
  "runtime-424bd133a78585db66ac.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "server.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`typescript-css-modules start should start a development server 1`] = `
SCRIPTS: [
  "/runtime.js",
  "/vendors-node_modules_pnpm_loadable_component_5_15_3_react_18_2_0_node_modules_loadable_compon-3dc5dd.js",
  "/main.js",
]
CSS: [
  "/main.css",
]
SOURCE HTML: <!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      My Awesome Project
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="stylesheet"
      href="styles[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[1]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[2]"
    >
  </head>
  <body>
    <div id="app">
      <div class="lessStyles__root___eRNyDeb">
        <div
          class="lessStyles__nested___MYunKdE"
          data-automation-text="true"
        >
          Hello World
        </div>
        <div>
          Render type:
          Initial
        </div>
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[0]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[1]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[2]"
    >
    </script>
  </body>
</html>
POST HYDRATE DIFFS: 
         </div>
         <div>
           Render type:
-          Initial
+          Client
         </div>
       </div>
     </div>

`;
